backend/src/db/migrations/20200312140328-bulk_upload_to_s3.js

```diff
@@ -0,0 +1,103 @@
+import { getDriver } from '../../db/neo4j'
+import { existsSync, createReadStream } from 'fs'
+import path from 'path'
+import { S3 } from 'aws-sdk'
+import mime from 'mime-types'
+import { s3Configs } from '../../config'
+
+export const description = `
+Upload all image files to a S3 compatible object storage in order to reduce
+load on our backend.
+`
+
+export async function up(next) {
+  const driver = getDriver()
+  const session = driver.session()
+  const transaction = session.beginTransaction()
+
+  const {
+    AWS_ENDPOINT: endpoint,
+    AWS_REGION: region,
+    AWS_BUCKET: Bucket,
+    S3_CONFIGURED,
+  } = s3Configs
+
+  if (!S3_CONFIGURED) {
+    // eslint-disable-next-line no-console
+    console.log('No S3 given, cannot upload image files')
+    return
+  }
+
+  const s3 = new S3({ region, endpoint })
+  try {
+    // Implement your migration here.
+    const { records } = await transaction.run('MATCH (image:Image) RETURN image.url as url')
+    let urls = records.map((r) => r.get('url'))
+    urls = urls.filter((url) => url.startsWith('/uploads'))
+    const locations = await Promise.all(
+      urls
+        .map((url) => {
+          return async () => {
+            const { pathname } = new URL(url, 'http://example.org')
+            const fileLocation = path.join(__dirname, `../../../public/${pathname}`)
+            const s3Location = `original${pathname}`
+            if (existsSync(fileLocation)) {
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-03-26T14:51:38Z" title="Thursday, March 26th 2020, 3:51:38 pm +01:00">Mar 26, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -32,39 +32,38 @@ export async function up(next) {
   try {
     // Implement your migration here.
     const { records } = await transaction.run('MATCH (image:Image) RETURN image.url as url')
-    let urls = records.map(r => r.get('url'))
-    urls = urls.filter(url => url.startsWith('/uploads'))
+    let urls = records.map((r) => r.get('url'))
+    urls = urls.filter((url) => url.startsWith('/uploads'))
     const locations = await Promise.all(
       urls
-        .map(url => {
+        .map((url) => {
           return async () => {
             const { pathname } = new URL(url, 'http://example.org')
             const fileLocation = path.join(__dirname, `../../../public/${pathname}`)
             const s3Location = `original${pathname}`
-            const mimeType = mime.lookup(fileLocation)
+            if (existsSync(fileLocation)) {
```


Can we maybe do sth. in `else`? E.g. delete the `url` from the Image node? Shall we do that in another migration?

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-03-26T15:15:16Z" title="Thursday, March 26th 2020, 4:15:16 pm +01:00">Mar 26, 2020</time>_

---


I thought about that, like we shouldn't have any `Image` nodes that point to files that don't exist in the `fileSystem`... like maybe we could have done that in the migration that created the `Image` nodes, but since that has already been deployed, we could iterate through all `Image`, check if they exist  in the file system and if not delete them and their relationship to `User/Post`, oder?